src/charm.py

```diff
@@ -358,6 +434,64 @@ def _set_secret(self, scope: str, key: str, value: Optional[str]) -> None:
         else:
             raise RuntimeError("Unknown secret scope.")
 
+    def _workload_initialise(self):
+        """Workload initialisation commands.
+
+        Create users and configuration to setup instance as an Group Replication node.
+        Raised errors must be treated on handlers.
+        """
+        self._mysql.configure_mysql_users()
+        self._mysql.configure_instance()
+        self._mysql.wait_until_mysql_connection()
+        workload_version = self._mysql.get_mysql_version()
+        self.unit.set_workload_version(workload_version)
+        self.unit_peer_data["instance-hostname"] = f"{instance_hostname()}:3306"
+
+    def _workload_reset(self, force: bool = False) -> StatusBase:
+        """Reset an errored workload.
+
+        Purge all files and re-initialise the workload.
+
+        Args:
+            force: flag to indicate if data needs to be forcefully purged
+
+        Returns:
+            A `StatusBase` to be set by the caller
+        """
+        # checks for data directory integrity
+        if not force and self._mysql.is_data_dir_initialised():
+            # avoid running for an apparently integral data dir
+            logger.error("Won't reset integral data directory.")
+            return BlockedStatus("data directory not empty")
+
+        try:
+            service_stop(SERVICE_NAME)
+            self._mysql.reset_data_dir()
+            self._mysql.reconfigure_mysqld()
+            self._workload_initialise()
+            unit_label = self.unit.name.replace("/", "-")
+            primary_address = self._get_primary_address_from_peers()
```

Comment:
may be we should use `self.mysql.get_cluster_primary_address()`?

Reply:
Commented above